src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -5,469 +5,253 @@ package app.revanced.patcher.apk
-import org.jf.dexlib2.Opcodes
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.MultiDexContainer
 import org.jf.dexlib2.writer.io.MemoryDataStore
 import org.w3c.dom.*
 import java.io.File
-import java.nio.file.FileSystems
-import java.nio.file.Files
-import kotlin.io.path.copyTo
+import java.util.zip.ZipEntry
+
 
 /**
- * The apk file that is to be patched.
+ * The [Apk] file that is to be patched.
  *
- * @param filePath The path to the apk file.
+ * @param path The path to the apk file.
+ * @param name The name of this apk.
  */
-sealed class Apk(filePath: String) {
-    /**
-     * The apk file.
-     */
-    open val file = File(filePath)
-
-    /**
-     * The patched resources for the [Apk] given by the [app.revanced.patcher.Patcher].
-     */
-    var resources: File? = null
-        internal set
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
 
-    /**
-     * The metadata of the [Apk].
-     */
-    val packageMetadata = PackageMetadata()
-
-    /**
-     * Get the resource directory of the apk file.
-     *
-     * @param options The patcher context to resolve the resource directory for the [Apk] file.
-     * @return The resource directory of the [Apk] file.
-     */
-    protected fun getResourceDirectory(options: PatcherOptions) = options.resourceDirectory.resolve(toString())
-
-    /**
-     * Get a file from the resources of the [Apk] file.
-     *
-     * @param path The path of the resource file.
-     * @param options The patcher context to resolve the resource directory for the [Apk] file.
-     * @return A [File] instance for the resource file.
-     */
-    internal fun getFile(path: String, options: PatcherOptions) = getResourceDirectory(options).resolve(path)
-
-    /**
-     * Transform public drawable resource references.
-     *
-     * @param options The patcher context to resolve the resource directory for the [Apk] file.
-     * @param transformer The callback with [Document] for public.xml to execute.
-     */
-    protected fun transformDrawableReferences(options: PatcherOptions, transformer: (Document) -> Unit) {
-        getResourceDirectory(options).resolve("res/values/public.xml").apply {
-            if (!exists()) return
+        private fun ApkModule.isFeatureModule() = androidManifestBlock.manifestElement.let {
+            it.searchAttributeByName("isFeatureSplit")?.valueAsBoolean == true || it.searchAttributeByName("configForSplit") != null
+        }
 
-            DomFileEditor(this).use { editor ->
-                transformer(editor.file)
+        fun new(path: File) = ApkModule.loadApkFile(path).let { module ->
+            when {
+                module.isBaseModule -> Base(module)
+                module.isFeatureModule() -> FeatureModule(module)
+                else -> Split(module)
             }
         }
```

Comment:
This should not be a method inside `Apk`. Instead get rid of the method and use `when` wherever you are calling `new`